src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,631 @@
+ * is chosen that supports the parameters passed to the {@code deriveKey} or
+ * {@code deriveData} method. However, if {@code getProviderName} or
+ * {@code getKDFParameters} is called before calling the {@code deriveKey} or
+ * {@code deriveData} methods, the first provider supporting the KDF algorithm
+ * and {@code KDFParameters} is chosen which may not be the provider that is
+ * eventually selected once the {@code AlgorithmParameterSpec} is supplied in
+ * the derive methods; therefore it is recommended not to call
+ * {@code getProviderName} or {@code getKDFParameters} until after a key
+ * derivation operation. Once a provider is selected, it cannot be changed.
+ * <p>
+ * API Usage Example:
+ * {@snippet lang = java:
+ *    KDF kdfHkdf = KDF.getInstance("HKDFWithHmacSHA256");
+ *
+ *    AlgorithmParameterSpec kdfParameterSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).thenExpand(info, 32);
+ *
+ *    kdfHkdf.deriveKey("AES", kdfParameterSpec);
+ *}
+ *
+ * @see KDFParameters
+ * @see SecretKey
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public final class KDF {
+    private static final Debug debug = Debug.getInstance("jca", "KDF");
+
+    private static final Debug pdebug = Debug.getInstance("provider",
+                                                          "Provider");
+    private static final boolean skipDebug = Debug.isOn("engine=")
+                                             && !Debug.isOn("kdf");
+
+    // The provider
+    private Provider provider;
+
+    // The provider implementation (delegate)
+    private KDFSpi spi;
+
+    // The name of the KDF algorithm.
+    private final String algorithm;
+
+    // Additional KDF configuration parameters
+    private final KDFParameters kdfParameters;
+
+    // next service to try in provider selection
+    // null once provider is selected
+    private Service firstService;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    private final Object lock;
+
+    int DERIVE_KEY = 0;
+    int DERIVE_DATA = 1;
+
+    /**
+     * Instantiates a KDF object.
+     *
+     * @param keyDerivSpi
+     *     the delegate
+     * @param provider
+     *     the provider
+     * @param algorithm
+     *     the algorithm
+     * @param kdfParameters
+     *     the algorithm parameters
+     */
+    private KDF(KDFSpi keyDerivSpi, Provider provider, String algorithm,
+                KDFParameters kdfParameters) {
+        this.spi = keyDerivSpi;
+        this.provider = provider;
+        this.algorithm = algorithm;
+        this.kdfParameters = kdfParameters;
+        firstService = null;
+        serviceIterator = null;
+        lock = new Object();
+    }
+
+    private KDF(Service s, Iterator<Service> t, String algorithm,
```

Comment:
nit: add comment that this constructor is used when no provider is supplied.

Reply:
Ok~

Reply:
I don't see the resolve this conversation setting though. Maybe it's up to the author?